platform/files.go

```diff
@@ -279,38 +283,79 @@ type RunImageMetadata struct {
 	Mirrors []string `toml:"mirrors" json:"mirrors,omitempty"`
 }
 
-func (rm *RunImageMetadata) BestRunImageMirror(registry string) (string, error) {
+//go:generate mockgen -package testmock -destination ../testmock/access_checker.go github.com/buildpacks/lifecycle/platform AccessChecker
+type AccessChecker interface {
+	CheckReadAccess(repo string, keychain authn.Keychain) (bool, error)
+}
+
+type AccessCheckerImpl struct{}
+
+var _ AccessChecker = &AccessCheckerImpl{}
+
+func (a *AccessCheckerImpl) CheckReadAccess(repo string, keychain authn.Keychain) (bool, error) {
+	img, err := remote.NewImage(repo, keychain)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse image reference")
+	}
+
+	return img.CheckReadAccess(), nil
+}
+
+func (rm *RunImageMetadata) BestRunImageMirror(registry string, accessChecker AccessChecker) (string, error) {
 	if rm.Image == "" {
 		return "", errors.New("missing run-image metadata")
 	}
+
 	runImageMirrors := []string{rm.Image}
 	runImageMirrors = append(runImageMirrors, rm.Mirrors...)
-	runImageRef, err := byRegistry(registry, runImageMirrors)
+
+	keychain, err := auth.DefaultKeychain(runImageMirrors...)
 	if err != nil {
-		return "", errors.Wrap(err, "failed to find run image")
+		return "", errors.Wrap(err, "unable to create keychain")
 	}
-	return runImageRef, nil
-}
 
-func (sm *StackMetadata) BestRunImageMirror(registry string) (string, error) {
-	return sm.RunImage.BestRunImageMirror(registry)
-}
+	runImageRef := byRegistry(registry, runImageMirrors, accessChecker, keychain)
+	if runImageRef != "" {
+		return runImageRef, nil
+	}
+
+	for _, image := range runImageMirrors {
+		ok, err := accessChecker.CheckReadAccess(image, keychain)
+		if err != nil {
+			return "", err
+		}
 
-func byRegistry(reg string, imgs []string) (string, error) {
-	if len(imgs) < 1 {
-		return "", errors.New("no images provided to search")
+		if ok {
+			return image, nil
+		}
 	}
 
-	for _, img := range imgs {
-		ref, err := name.ParseReference(img, name.WeakValidation)
+	return "", errors.New("failed to find accessible run image")
+}
+
+func (sm *StackMetadata) BestRunImageMirror(registry string, accessChecker AccessChecker) (string, error) {
+	return sm.RunImage.BestRunImageMirror(registry, accessChecker)
+}
+
+func byRegistry(reg string, repos []string, accessChecker AccessChecker, keychain authn.Keychain) string {
+	for _, repo := range repos {
+		ref, err := name.ParseReference(repo, name.WeakValidation)
 		if err != nil {
 			continue
 		}
 		if reg == ref.Context().RegistryStr() {
-			return img, nil
+			ok, err := accessChecker.CheckReadAccess(repo, keychain)
+			if err != nil {
+				return ""
```

Comment:
If an error occurs at `accessChecker.CheckReadAccess` - how will we be able to see the failure and get the information to the user?

Reply:
Currently - not at all. The next working mirror would be taken and if no run image could be found a generic error would be risen. The problem is that there might be several errors if there are several mirrors. Or even some fail and others works.

Reply:
It might make sense to log a warning here

Reply:
> It might make sense to log a warning here

2 quick follow-up questions:

1. If a run-image matching the image's registry was found without read access, a warning sounds reasonable. Do you think we should also warn for "normal" run-mirrors? For example, for use one mirror would be located in the VPN, so this warning would be kind of normal when you are not in the VPN.
2. Is there a better way to pass around the `log.Logger` in each function call?

Reply:
> Do you think we should also warn for "normal" run-mirrors?

You're right, that might get noisy. Let's not warn in this case.

Re: passing around the logger, I'm not sure, we're not very consistent about this (sometimes the logger is embedded, sometimes it's just passed around). I'm happy to defer to whatever you think is best.

Reply:
> I'm happy to defer to whatever you think is best.

We had a closer look and the changes needed to pass around the logger seem not to be worth logging a warning in this edge case. This would be if the run image matches the registry of the image, the image can be written, but the run image can't be pulled. Right? WDYT?